Update the fsl package recipe

glennpj opened the pull request:
- added several patches
- added some missing dependencies
- remove unneeded dependencies
- add CUDA support
- disable queue support, which was limited, and broken anyway
- move package text that was specific to the package to a comment, so it
  does not show up the environment module
- set conflicts for cuda and compilers

glennpj committed:
Update the fsl package recipe

- added several patches
- added some missing dependencies
- remove unneeded dependencies
- add CUDA support
- disable queue support, which was limited, and broken anyway
- move package text that was specific to the package to a comment, so it
  does not show up the environment module
- set conflicts for cuda and compilers

glennpj commented:
Can someone take a look at this?

Thanks.

alalazo commented on var/spack/repos/builtin/packages/fsl/package.py:
```diff
@@ -8,49 +8,163 @@
+    conflicts('platform=darwin',
+              msg='currently only packaged for linux')
+
+    patch('build_log.patch')
+    patch('eddy_Makefile.patch', when='@6:')
+    patch('fslpython_install_v5.patch', when='@:5')
+    patch('fslpython_install_v6.patch', when='@6:')
+    patch('fsl_sub_v5.patch', when='@:5')
+    patch('fsl_sub_v6.patch', when='@6:')
+    patch('iconv.patch')
```
Can you add a comment saying where the patches are coming from?

glennpj replied:
I created the patches. I am not sure how you want that indicated.

adamjstewart replied:
Have these patches been contributed upstream?

glennpj replied:
I normally do contribute upstream. However, the distribution of FSL requires registration and so it is not clear to me where patches would be submitted. Also, most of these patches would not apply upstream anyway.

The following patches are specific to spack:
- build_log.patch
- fslpython_install_v6.patch

The following is only necessary for the older version so is no longer relevant upstream.
- fslpython_install_v5.patch

The following are special cases. I referenced this in my initial commit but I will add a comment to the package recipe for this to be clearer.
- fsl_sub_v5.patch
- fsl_sub_v6.patch
These patches disable FSL's attempts to try to submit a subset of FSL computations to an SGE queue system. That auto-submit mechanism only works for SGE and requires someone to edit the `fsl_sub` script to accommodate their system. These patches disable the auto submission scheme and allow the `fsl_sub` script to behave the same on all systems, and without further modification, whether the computation is submitted to a "local" system, like a workstation, or as a batch job to a cluster queueing system, regardless of queue system type. In other words, all FSL computations would be submitted as a job script created by the user, as opposed to some computations attempting to auto-submit while other computations require a user prepared submission script, which winds up being confusing to users. These patches make no attempt to "fix" the behavior to make it work on other queueing systems, etc., and thus would be of no interest upstream.

If there is disagreement to having the fsl_sub patches as I have done then perhaps a variant is an option. I will also add that FSL needs a conda environment to be installed after installation, which is kind of ugly. Because of that, although we use FSL at our site, I probably would not have submitted this as a PR if there was not an FSL package already in spack.

alalazo commented on var/spack/repos/builtin/packages/fsl/package.py:
```diff
@@ -8,49 +8,163 @@
+        vtk_settings.filter(r'(^VTKSUFFIX)\s*=.*',
+                            r'\1 = -{0}'.format(vtk_suffix))
+
+        if 'cuda' in self.spec:
```
```suggestion
        if '+cuda' in self.spec:
```

alalazo commented on var/spack/repos/builtin/packages/fsl/package.py:
```diff
@@ -8,49 +8,163 @@
 
+    @run_after('install')
+    def postinstall(self):
+        del os.environ['PYTHONPATH']
+        del os.environ['PYTHONHOME']
```
Can you put this within the:
```python
with spack.util.environment.preserve_environment('PYTHONPATH', 'PYTHONHOME'):
    ....
```
context manager? Asking to ensure that the environment is restores after executing the script.

glennpj replied:
Okay; I am not really familiar with that. The only other package that I see with that is octave so I assume that is a good example.

alalazo reviewed (changes requested):
A few minor request.

glennpj committed:
Update var/spack/repos/builtin/packages/fsl/package.py

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

glennpj committed:
Delete python variables in the preserve_environment context manager

glennpj committed:
Add explanatory comment for fsl_sub patches

alalazo commented on var/spack/repos/builtin/packages/fsl/package.py:
```diff
@@ -8,49 +8,178 @@
+            for v in vars_to_unset:
+                del os.environ[v]
+
+        script = Executable(join_path(prefix, 'etc', 'fslconf',
+                                      'post_install.sh'))
+        script('-f', prefix)
```
I think these two lines need to be indented one level more: the variables are deleted only within the context manager, but will be restored outside of it.

glennpj replied:
Yes, sorry, I meant to do that but missed it as I was trying to understand the context manager. I will push out the fix.

glennpj committed:
Indent the script resources under context manager

The `script` resources need to be indented under the
`preserve_environment` context manager.
